core/Segment.php

```diff
@@ -460,6 +464,32 @@ public function getHash()
 
     public static function getSegmentHash($definition)
     {
+        $cache = Cache::getEagerCache();
+        $cacheKey = self::CACHE_KEY . md5($definition);
+
+        if ($cache->contains($cacheKey)) {
+            return $cache->fetch($cacheKey);
+        }
```

Comment:
I cache segments, can you please check @diosmosis if this is a good way to use the cache? I used eager cache, because every request calls this, even if we request all visits.

Reply:
Some matomo users create a very large amount of segments, so loading these on every request might be an issue. I'm also not sure if the tracker uses the eager cache, but it seems possible to me that it could hit some code that uses it and loads all this data. If the issue is to avoid hitting the database, then using the Lazy cache should be ok? It would just hit the filesystem. 

Does this code also get executed if `$definition == ''`? There could be a special check for empty string if so.

Can you list some of the cases where this is called even when there is no segment supplied? (For example, when `$definition == ''`, etc.) Does it happen only on API requests to get data or on requests, eg, in the admin section and elsewhere?

cc @tsteur in case he has a different opinion.

Reply:
I'd probably use a lazy cache here simply because it shouldn't be needed in most requests (assuming the method is not called for an empty definition string or maybe for an empty definition string we could always return same hard coded `return md5('');`? ). Assuming this changes rarely we could cache them for quite a long time.

If the segment hash can ever change then we'd need to invalidate the cache on update. 

Initially I thought to cache only the list of definitions and their hashes and then iterate over them every time (like `[['definition' => '...', 'hash' => '...'], [...], [...]]`) to find the right hashes so we don't create too many cache files if there are heaps of segments (as it can slow down the filesystem access) but then of course the downside is needing to load a larger cache file as it includes the information for all caches. I'd probably rather go with this kind of catch and match a segment by looping over the arrays but either way might be fine. Both have their benefits and disadvantages and it's not clear which one is better.

Something to still look at is how to handle the idSites and when the same definition was to exist for different sites with different hashes. This one may be super tricky. It wouldn't be too important right now as they technically should have all the same hashes for the same definition but this might not always be the case due to bugs in the past and if the calculation in the future changes.

Reply:
We call it a LOT. On a page reload (eg on the devices report), without choosing any segment, there are 300 calls for that method. Then click on another report (ajax calls), we have more than 200 calls again.
Currently, it does not resource-intensive, because we don't hit the DB, but we will with the new version.
If we go with the lazy cache, @tsteur list idea is good, because that way we don't need multiple files.

@diosmosis we have a check for the empty definition, but it looks like some plugins load all the segments even if the definition is empty (we call the `getHash` even more, but if the `definition` is empty, it doesn't call the other method).

Reply:
I tested again, how many times we call the function. 
The tracker didn't call it.
The dashboard is the worst. This time it called it more than 2000 times when browser archiving was on, and around 1000 times when it was off.
Other pages call it less. 

But it's much more if I choose any segments. The numbers above were with all visits.

I checked the eager cache file for the demo site. Currently, it's 530K. When I added 100 segment hashes (by hand, locally) it went up to 539K. I believe that won't be an issue.

Reply:
(Answering because I don't see a reply from @tsteur. If you've already discussed it w/ him, feel free to ignore my comment.)

You're right it doesn't seem like adding segment hashes will cause the eager cache to get significantly larger. Unless there's, say, a matomo instance w/ 1000 users and each one for some reason creates an average of 50 segments (which isn't entirely implausible since I've seen segments like `pageUrl==...,pageUrl==...,pageUrl==...,...` for like maybe 50 different page URLs). In this case we'd add potentially 50000 hashes to the eager cache.

I also believe you when you say the tracker didn't call it or use the eager cache. The tracker and the rest of matomo, however, aren't strictly separated, so it would be pretty easy for some change in the future to accidentally trigger a code path that uses the eager cache, even without ever using the `Segment` class. Maybe in the case of the hyopthetical user above, this happens, and tracking slows down a bit. Since tracking is meant to handle a high volume of concurrent requests, any performance degradation can be seen as a problem.

Both of these situations are hypothetical and would not be common if they happened at all, but edge cases always happen in some form, and this one would be particularly annoying, since the only sign of an issue we'd have is "tracking is a little slower". And all of these edge cases could be avoided by just using the Lazy cache. Which, to my knowledge, would not be noticeably slower than if we used the eager cache here.

I also believe you when you say the segment hashing method is called many times on every request. However, that's clearly a surprise, and not something we expect or want. So ideally we'd get some more context as to why that happens, where it happens, and what we can do about it, rather than optimize around something that we don't to happen in the first place.

All of that said, I am not the authority here, so @tsteur can of course give his opinion as well.

Reply:
thanks for the comment @diosmosis.
We had a chat with @tsteur this morning. I'm pretty sure lots of edge cases can happen, and I can't see through the codebase as much as you, so every comment and recommendation is very helpful.

I modified the caching and added comments to the code. Can you please both check it and let me know what you think about it?

Another idea I have: currently we use the hashing in a couple of places, so probably the `createHash` method in the `SegmentEditor` model should be public static, so we call that everywhere, this way when we modify it, we just modify in one place.